src/rules.rs

```diff
@@ -437,6 +450,29 @@ impl TraitDatum {
         }).cast();
 
         let mut clauses = vec![wf];
+
+        if !self.binders.value.flags.external {
+            let impl_allowed = self.binders.map_ref(|bound_datum|
+                ProgramClauseImplication {
+                    consequence: DomainGoal::LocalImplAllowed(bound_datum.trait_ref.clone()),
+                    conditions: Vec::new(),
+                }
+            ).cast();
+
+            clauses.push(impl_allowed);
+        } else {
+            let impl_maybe_allowed = self.binders.map_ref(|bound_datum|
+                ProgramClauseImplication {
+                    consequence: DomainGoal::LocalImplAllowed(bound_datum.trait_ref.clone()),
+                    conditions: vec![
+                        DomainGoal::IsLocal(bound_datum.trait_ref.parameters[0].assert_ty_ref().clone()).cast(),
+                    ],
+                }
+            ).cast();
+
+            clauses.push(impl_maybe_allowed);
+        }
```

Comment:
nit: These `if` branches have some code duplication, you could DRY it up by doing:
```rust
let mut conditions = Vec::new();
if !self.binders.value.flags.external {
    conditions.push(DomainGoal::IsLocal(....).cast());
}
// ...make ProgramClause and push it
```